File: Data Structures & Algorithms/sqrtx/submission-6.java
class Solution {
    public int mySqrt(int x) {
        if (x == 0) return 0;

        int left = 1;
        int right = x;
        int res = 1;

        while (left <= right) {
            int middle = left + (right - left) / 2;
            long squared = (long) middle * middle;

            if (squared == x) return middle;

            if (squared > x) {
                right = middle - 1;
            } else {
                res = middle;
                left = middle + 1;
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-6.ts
class Solution {
    /**
     * @param {number[]} piles
     * @param {number} h
     * @return {number}
     */
    minEatingSpeed(piles: number[], h: number): number {
        // Two pointers that represent the range of possible bananas / hour eaten.
        let left: number = 1;
        let right: number = Math.max(...piles);
        let res = right;

        while (left <= right) {
            const k: number = Math.floor((left + right) / 2);
            let totalTime = 0;

            for (const pile of piles) {
                totalTime += Math.ceil(pile / k);
            }   

            if (totalTime <= h) {
                res = k;
                right = k - 1;
            } else {
                left = k + 1;
            }
        }

        return res;
    }
}
